xalia/Win32/HwndProvider.cs

```diff
@@ -113,16 +125,95 @@ public Task<(bool, int, int)> GetClickablePointAsync(UiDomElement element)
 
         public string[] GetTrackedProperties()
         {
-            return null;
+            return tracked_properties;
         }
 
         public void NotifyElementRemoved(UiDomElement element)
         {
             Element = null;
         }
 
+        private void WatchChildren()
+        {
+            if (_watchingChildren)
+                return;
+            if (Element.MatchesDebugCondition())
+                Utils.DebugWriteLine($"WatchChildren for {Element} (win32)");
+            _watchingChildren = true;
+            Utils.RunIdle(PollChildren); // need to wait for any other providers to remove their children
+        }
+
+        private void PollChildren()
+        {
+            if (!_watchingChildren)
+                return;
+
+            var child_hwnds = EnumImmediateChildWindows(Hwnd);
+
+            // First remove any existing children that are missing or out of order
+            int i = 0;
+            foreach (var new_child in child_hwnds)
+            {
+                var child_name = $"hwnd-{new_child}";
+                if (!Element.Children.Exists((UiDomElement element) => element.DebugId == child_name))
+                    continue;
+                while (Element.Children[i].DebugId != child_name)
+                {
+                    Element.RemoveChild(i);
+                    _childCount--;
+                }
+                i++;
+            }
+
+            // Remove any remaining missing children
+            while (i < _childCount)
+            {
+                Element.RemoveChild(i);
+                _childCount--;
+            }
+
+            // Add any new children
+            i = 0;
+            foreach (var new_child in child_hwnds)
+            {
+                var child_name = $"hwnd-{new_child}";
+                if (_childCount <= i || Element.Children[i].DebugId != child_name)
+                {
+                    if (!(Connection.LookupElement(child_name) is null))
+                    {
+                        // Child element is a duplicate of another element somewhere in the tree.
+                        continue;
+                    }
+                    Element.AddChild(i, Connection.CreateElementFromHwnd(new_child));
+                    _childCount++;
+                }
+                i += 1;
+            }

```

Comment:
It kind of feels confusing to me to be checking all of the children of Element when we're just working with children of a particular provider. I feel like maybe it'd make more sense to maintain a separate list for children of each individual provider, and maybe have the order of children on the main element be children of provider one, children of provider 2, and so on.

Reply:
I thought about that, but it seems like a lot of work to aggregate each provider's children and keep track of them. I'm also a little fussy about the exact order - specifically I want "arbitrary" child elements like child hwnds, IAccessible's, etc, to be before one-off things like the win32 vertical scrollbar, which would likely conflict with the order you would get when you enforce a specific order of each provider's children.

Reply:
So long as the assumption is that HwndProvider children are always the first children of the entire element, this holds up fine and although it confuses me a bit initially I can't really think of any ways to improve it necessarily. But, if that ever changes, i.e you start having children that aren't HwndProvider children at index 0, I don't think this will work.

Reply:
Well, unless you choose to skip over child elements that come from another provider I guess. So you'd maintain a consistent order of children of this provider, while not removing children that come from other providers.

Reply:
Yes, my plan was that only one provider (specifically the one that understands the current value of `recurse_method`), can provide a set of child elements starting at index 0, and all others should add child elements at the end and not worry about their order. I care about this because the indices of children in an external tree (such as the tree of hwnds or IAccessible objects) can be semantically meaningful. In particular, we need them for determining the selected tab of tab controls in Qt on Linux.